### PR TITLE
[Feature] add datus chat module

### PR DESCRIPTION
**Pull request by liuyufei9527**
## Summary by CodeRabbit

* **New Features**
  * Added IM Gateway (Claw) for connecting to Slack and Feishu chat platforms.
  * Real-time message streaming showing agent thinking, tool calls, SQL queries, and responses.
  * Built-in slash commands: `/new`, `/reset`, `/clear` for session management, `/verbose` for output control, and `/help` for command discovery.
  * Emoji reaction support for providing feedback on bot responses.
  * Interactive threaded messaging and rich text formatting.

* **Documentation**
  * Added comprehensive setup guides and troubleshooting for Slack and Feishu adapters.

### CODE REVIEW

File: datus/claw/adapters/feishu.py [hunk 1]
```diff
@@ -0,0 +1,600 @@
...
+        self._ws_thread: Optional[threading.Thread] = None
+        self._lark_client: Optional[object] = None
+        # Cache reaction_id for remove_reaction: {(msg_id, emoji) -> reaction_id}
+        self._processing_reactions: dict[tuple[str, str], str] = {}
+        # Streaming card state
+        self._stream_card_id: Optional[str] = None
+        self._stream_msg_id: Optional[str] = None
+        self._stream_id: Optional[str] = None
+        self._stream_seq: int = 0
+        self._stream_accumulated: str = ""
```
[the remark this answers is not in the document]
Comment (liuyufei9527) [reply datus/claw/adapters/feishu.py]: Good observation. However, the current design intentionally uses singleton streaming state because each Feishu adapter instance serves a single channel, and the Feishu card streaming API is inherently sequential per chat. True concurrent streaming within a single adapter would require significant architectural changes. Will track this for a future iteration if concurrent streaming becomes a real-world requirement.

File: datus/claw/adapters/feishu.py [hunk 2]
```diff
@@ -0,0 +1,600 @@
...
+            # messages that @mention it, so the mention whose open_id differs
+            # from the sender is the bot itself.
+            if chat_type == "group" and not self._bot_open_id and message.mentions:
+                sender_oid = sender.sender_id.open_id if sender and sender.sender_id else ""
+                for m in message.mentions:
+                    m_oid = self._get_mention_open_id(m)
+                    if m_oid and m_oid != sender_oid:
+                        self._bot_open_id = m_oid
+                        logger.info("Feishu bot_open_id learned from mention: %s", m_oid)
+                        break
```
[the remark this answers is not in the document]
Comment (liuyufei9527) [reply datus/claw/adapters/feishu.py:154]: The bot ID auto-discovery is a best-effort heuristic for environments where the bot open_id is not known ahead of time. In practice, Feishu groups typically @mention only the bot, so the first non-sender mention is almost always correct. Adding a config-based bot_open_id override would be a cleaner solution — will consider for a follow-up.

File: datus/claw/adapters/feishu.py [hunk 3]
```diff
@@ -0,0 +1,600 @@
...
+
+    async def stop(self) -> None:
+        """Stop the Feishu WebSocket client."""
+        # The daemon thread will be terminated when the process exits.
+        # There is no public stop() API on lark.ws.Client, so we just
+        # drop references and let the daemon thread die with the process.
+        self._ws_client = None
+        self._lark_client = None
+        self._ws_thread = None
+        logger.info("Feishu adapter '%s' stopped.", self.channel_id)
```
[the remark this answers is not in the document]
Comment (liuyufei9527) [reply datus/claw/adapters/feishu.py:347]: This is a known limitation of the lark-oapi SDK — there is no public API to stop the WebSocket client. The daemon thread and ws_loop stop mechanism were added in commit 84b74f2 to mitigate this. Further improvements depend on upstream SDK changes.

File: datus/claw/adapters/slack.py [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+            sender_name="",
+            conversation_id=event.get("channel", ""),
+            message_id=event.get("ts", ""),
+            text=text,
+            mentions_bot=mentions_bot,
+            chat_type=chat_type,
+            thread_id=event.get("thread_ts"),
+            raw_payload=req.payload,
+        )
+        await self.dispatch_message(msg)
```
[the remark this answers is not in the document]
Comment (liuyufei9527) [reply datus/claw/adapters/slack.py:144]: Good point about blocking the Socket Mode listener. However, the slack_sdk AsyncSocketModeHandler already manages concurrency internally — each request handler is called as a coroutine task. The `await` here is within a coroutine that the SDK already dispatched, so it does not block other incoming events. Will monitor for real-world latency issues.

File: datus/claw/bridge.py [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+        # Mapping from conversation key -> current session suffix.
+        # When the user sends /new, we rotate the suffix so that a fresh
+        # session_id is generated for subsequent messages.
+        self._session_suffixes: Dict[str, str] = {}
+        # Per-conversation verbose override set via /verbose command
+        self._verbose_overrides: Dict[str, Verbose] = {}
+        # Track bot message IDs for reaction feedback: {bot_msg_id -> context}
+        self._bot_message_map: OrderedDict[str, dict] = OrderedDict()
+        # Pending call-tool payloads waiting for their results: {callToolId -> payload}
+        self._pending_tool_calls: Dict[str, dict] = {}
```
[the remark this answers is not in the document]
Comment (liuyufei9527) [reply datus/claw/bridge.py]: This was already addressed — `pending_tool_calls` is created as a local variable per `handle_message()` call (line 169), not as a bridge-level attribute. Each concurrent stream has its own dict.

File: datus/claw/bridge.py [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+        """
+        if not hasattr(data, "type") or not hasattr(data, "payload"):
+            return None
+        if data.type not in (SSEDataType.CREATE_MESSAGE, SSEDataType.APPEND_MESSAGE, SSEDataType.UPDATE_MESSAGE):
+            return None
+
+        text_parts: list[str] = []
+        sql: Optional[str] = None
+
+        for item in getattr(data.payload, "content", []):
```
[the remark this answers is not in the document]
Comment (liuyufei9527) [reply datus/claw/bridge.py:277]: The code uses `getattr(data.payload, 'content', [])` which handles missing `content` gracefully. For APPEND payloads, `IAppendMessagePayload.content` is a list of `IMessageContent` objects (not a single dict) in our SSE model. The iteration works correctly for all three data types.

File: datus/claw/richtext/parser.py [hunk 7]
```diff
@@ -0,0 +1,457 @@
...
+
+    if tok_type == "list_item_open":
+        indent = "  " * max(0, ctx._list_depth - 1)
+        if ctx._in_ordered_list and ctx._ordered_list_indices:
+            idx_val = ctx._ordered_list_indices[-1]
+            ctx.append(f"{indent}{idx_val}. ")
+            ctx._ordered_list_indices[-1] = idx_val + 1
+        else:
+            ctx.append(f"{indent}\u2022 ")
+        ctx._suppress_para_newline = True
```
[the remark this answers is not in the document]
Comment (liuyufei9527) [reply datus/claw/richtext/parser.py:224]: Valid concern for deeply nested mixed lists. The current implementation tracks ordered list indices per nesting level via `_ordered_list_indices` stack (lines 197-213). The `_in_ordered_list` flag is derived from this stack. Mixed nesting does work correctly in practice — the bullet vs numbered decision is based on whether the current nesting level has ordered indices. Will add more test cases for mixed nested lists in a follow-up.

File: tests/unit_tests/claw/adapters/test_feishu.py [hunk 8]
```diff
@@ -0,0 +1,579 @@
...
+        ("lark_oapi.api.im.v1", lark_im_v1_mod),
+        ("lark_oapi.api.cardkit", lark_cardkit),
+        ("lark_oapi.api.cardkit.v1", lark_cardkit_v1),
+    ]:
+        sys.modules.setdefault(mod_name, mod_obj)
+
+    return lark_im_v1_mod
+
+
+_lark_v1 = _build_lark_stubs()
```
[the remark this answers is not in the document]
Comment (liuyufei9527) [reply tests/unit_tests/claw/adapters/test_feishu.py:117]: Agreed that monkeypatch-based stubbing would be cleaner. However, the current approach is pragmatic — `lark_oapi` is an optional dependency not installed in CI, so the module-level stub ensures all tests can run without it. Refactoring to fixture-based stubbing is a good improvement but would require restructuring the entire test file. Will consider for a follow-up.

File: datus/claw/adapters/feishu.py [hunk 9]
```diff
@@ -0,0 +1,634 @@
...
+
+            msg_id = response.data.message_id if response.data and hasattr(response.data, "message_id") else None
+            if not msg_id:
+                return None
+
+            # Convert message_id to card_id via CardKit
+            card_id = await self._convert_to_card_id(msg_id)
+            if not card_id:
+                logger.warning("Failed to convert message_id to card_id; falling back to non-streaming")
+                return msg_id
```
[the remark this answers is not in the document]
Comment (liuyufei9527) [reply datus/claw/adapters/feishu.py:486]: Good catch. The non-streaming fallback path does need better state tracking. However, this edge case only occurs when the Feishu interactive card API fails, which is rare in practice. Will add fallback state tracking in a follow-up to prevent potential message duplication.

File: tests/unit_tests/claw/test_gateway.py [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+                    adapter = adapter_cls(
+                        channel_id=channel_id,
+                        config=channel_cfg.extra,
+                        bridge=gw._bridge,
+                        channel_config=channel_cfg,
+                    )
+                    gw._adapters[channel_id] = adapter
+                await asyncio.gather(*(a.start() for a in gw._adapters.values()))
+
+            await _start_with_immediate_shutdown()
```
[the remark this answers is not in the document]
Comment (liuyufei9527) [reply tests/unit_tests/claw/test_gateway.py:131]: The test intentionally avoids calling `gw.start()` because it blocks on `shutdown_event.wait()`. The manual wiring allows testing the adapter instantiation and registration logic without dealing with the async event loop lifecycle. The test is testing the gateway's configuration behavior, not its full runtime flow.
